src/jaxsim/api/ode.py

```diff
@@ -150,14 +150,14 @@ def system_velocity_dynamics(
         # Sum the forces of all collidable points rigidly attached to a body.
         # Since the contact forces W_f_Ci are expressed in the world frame,
         # we don't need any coordinate transformation.
-        W_f_Li_terrain = jax.vmap(
-            lambda nc: (
-                jnp.vstack(
-                    jnp.equal(parent_link_index_of_collidable_points, nc).astype(int)
-                )
-                * W_f_Ci
-            ).sum(axis=0)
-        )(jnp.arange(model.number_of_links()))
+        W_f_Li_terrain = jnp.where(
+            (
+                parent_link_index_of_collidable_points[:, None]
+                == jnp.arange(model.number_of_links())
+            ).any(axis=-1, keepdims=True),
+            W_f_Ci,
+            jnp.zeros_like(W_f_Ci),
+        ).sum(axis=0)
```

Comment:
The idea was both to make the code simpler, as the current version is quite intricate, and reduce the memory footprint of that part of code. If you prefer, I can split the binary mask and the `jnp.where(...)` as:

```suggestion
        mask = (
            parent_link_index_of_collidable_points[:, None]
            == jnp.arange(model.number_of_links())
        ).any(axis=-1, keepdims=True)

        W_f_Li_terrain = jnp.where(
            mask,
            W_f_Ci,
            jnp.zeros_like(W_f_Ci),
        ).sum(axis=0)
```